New addon: `flag-menu`

DNin01 opened the pull request:
Resolves #7230

### Changes

- Adds a new addon which adds a context menu to the green flag. It lets you toggle turbo mode, higher framerate, and mute. Now mobile users can finally use turbo mode without having to enter the editor!
- Related addons have been updated to utilize their context menu items.
- Also fixed an issue that prevents the displayNoneWhileDisabled API from working on addons beginning with a number.

It looks like this:
![Green flag context menu](https://github.com/user-attachments/assets/f75821ed-3434-4506-81f9-c66b8b027f5d)

*Context menu items (other than toggle turbo mode) only appear if their respective addons are enabled.

### Tests

Tested on Edge 140 with a mouse and with DevTools Device Emulation.

To do:
- [ ] Use `touchstart` instead of `contextmenu`
- [x] Localize
- [x] Dynamic disable, enable, and setting changes work for all updated addons
- [x] `editor-dark-mode` compatibility
- [x] `editor-compact` compatibility

DNin01 committed:
New addon: `flag-menu`

DNin01 committed:
Format code

DNin01 committed:
Localize addon

DNin01 committed:
Apply color and compact styles

mxmou commented on addons/60fps/addon.json:
A notice explaining how to access 60 FPS mode on a touch screen would be helpful. Same with mute-project.

DNin01 replied:
That addon already has so much notice text. 🤣

WorldLanguages replied:
If you want to add `"touchscreenOnly": true"` to notices within this PR it's fine to me.

mxmou commented on addons-l10n/en/flag-menu.json:
```diff
@@ -0,0 +1,5 @@
+{
+  "flag-menu/toggle-turbo": "toggle turbo mode",
```
Separate messages for turning it on and off would be more consistent with Scratch's UI and easier to translate into other languages. Similarly, "mute" and "unmute" would be better than "toggle mute".

DNin01 replied:
That was my original plan—I will add it later.

mxmou commented on addons/flag-menu/context_menu.css:
```diff
@@ -0,0 +1,42 @@
+  display: none;
+}
+
+/* Styles reused from https://github.com/scratchfoundation/scratch-gui/blob/develop/src/components/context-menu/context-menu.css */
```
These styles will soon be outdated because Scratch's [React 18 PR](https://github.com/scratchfoundation/scratch-editor/pull/299) changes context menus. Can you use `addon.tab.scratchClass` so that the addon doesn't need to duplicate Scratch CSS?

DNin01 replied:
Alright. It seems to work fine.

mxmou reviewed (changes requested):
This addon should be enabled by default - otherwise touchscreen users can't use 60fps and mute-project at all without enabling it first.

After the React 18 update, it will be possible to navigate Scratch's context menus using arrow keys. It would be nice if the addon replicated that.

DNin01 committed:
Clone Scratch classes

DNin01 commented:
> This addon should be enabled by default.

Perhaps enabled by default only for mobile versions of the extension?

That's too much work, though. I'll just make it enabled by default.

> After the React 18 update, it will be possible to navigate Scratch's context menus using arrow keys. It would be nice if the addon replicated that.

Ideally, I would make the context menu an instance of a React context menu so that React can take care of this. Not sure how, though. May just have to stick with custom handlers.

We don't use React within our project, do we? (We probably should.)

DNin01 committed:
toggle mute -> mute/unmute

DNin01 committed:
Remove default mute label text

DNin01 committed:
toggle -> turn on/off

DNin01 commented:
Okay, I got all those kinks worked out!

DNin01 committed:
Enable by default

mxmou commented:
> Ideally, I would make the context menu an instance of a React context menu so that React can take care of this. Not sure how, though. May just have to stick with custom handlers.

Writing the event handlers is probably easier, especially because we currently need to support both versions of React. [That's what I did when updating the context menu API.](https://github.com/ScratchAddons/ScratchAddons/blob/master/addon-api/content-script/contextmenu.js#L19-L53) After the React 18 update is out, we can experiment with using Scratch's context menu component.

Speaking of React 18: if you don't mind, I'll push some changes to make the addon's styling work there.

DNin01 commented:
> Speaking of React 18: if you don't mind, I'll push some changes to make the addon's styling work there.

Sure, go ahead.

mxmou committed:
React 18 support

mxmou committed:
Format code

DNin01 commented on addons/flag-menu/userscript.js:
```diff
@@ -0,0 +1,83 @@
+  let greenFlag;
+
+  const contextMenuClass =
+    addon.tab.scratchClass("context-menu_context-menu") || addon.tab.scratchClass("context-menu_context-menu-content"); // React 16 || React 18
```
Scratch updated React to 18 with the Face Sensing update, so this can be simplified. The style of the context menu was also changed, so some things may need to be tweaked.

DNin01 replied:
I mean, Scratch's new context menu has no transition for fading in/out, it just appears immediately, plus there's arrow-key navigation and the width was reduced. Every other property applies automatically.

mxmou committed:
Simplify React 18 code

The React 18 update has been released, so we no longer need to support React 16.

DNin01 committed:
Simplify function parameters

DNin01 commented:
<details><summary>Still working on an implementation that skips over invisible menu items, but here's what I have now. I think it's alright.</summary>

```diff
  export default async function ({ addon, console, msg }) {
    let greenFlag;
+   let contextMenuOpen = false;

    const contextMenu = Object.assign(document.createElement("nav"), {
      role: "menu",
      className: addon.tab.scratchClass("context-menu_context-menu-content", { others: "sa-flag-context-menu" }),
    });
+   const setHighlightedItem = (menuitem) => {
+     const currentItem = contextMenu.querySelector("[data-highlighted]");
+     if (currentItem) {
+       currentItem.removeAttribute("data-highlighted");
+       currentItem.tabIndex = -1;
+     }
+     if (menuitem) {
+       menuitem.setAttribute("data-highlighted", "");
+       menuitem.tabIndex = 0;
+       menuitem.focus();
+     }
+   }
    const createItem = (id, textContent = "") => {
      const item = Object.assign(document.createElement("div"), {
        role: "menuitem",
        className: addon.tab.scratchClass("context-menu_menu-item"),
        id,
        textContent,
      });
-     item.addEventListener("mouseenter", () => item.setAttribute("data-highlighted", ""));
-     item.addEventListener("mouseleave", () => item.removeAttribute("data-highlighted"));
+     item.addEventListener("mouseenter", () => setHighlightedItem(item));
+     item.addEventListener("mouseleave", () => setHighlightedItem());
      return item;
    };
    contextMenu.append(
      createItem("sa-flag-menu-turbo", msg("turbo-on")),
      createItem("sa-flag-menu-fps"),
      createItem("sa-flag-menu-mute")
    );

    function closeContextMenu() {
+     contextMenuOpen = false;
+     setHighlightedItem();
      contextMenu.classList.remove("sa-flag-menu-open");
    }
    document.addEventListener("mousedown", (e) => {
      if (!e.target.closest("[role='menu']")) closeContextMenu();
    });
    document.addEventListener("keydown", (e) => {
-     if (e.key === "Escape") closeContextMenu();
+     if (!contextMenuOpen) return;
+     switch (e.key) {
+       case "Escape":
+       case "Tab":
+         closeContextMenu();
+       break;
+       case "Enter":
+       case " ":
+         if (contextMenu.querySelector("[data-highlighted]")) contextMenu.querySelector("[data-highlighted]").click();
+       break;
+       case "ArrowUp":
+         const previousItem = contextMenu.querySelector("[data-highlighted]")?.previousElementSibling ?? contextMenu.lastElementChild;
+         setHighlightedItem(previousItem, true);
+       break;
+       case "ArrowDown":
+         const nextItem = contextMenu.querySelector("[data-highlighted]")?.nextElementSibling ?? contextMenu.firstElementChild;
+         setHighlightedItem(nextItem, true);
+       break;
+       case "Home":
+       case "PageUp":
+         setHighlightedItem(contextMenu.firstElementChild, true);
+       break;
+       case "End":
+       case "PageDown":
+         setHighlightedItem(contextMenu.lastElementChild, true);
+       break;
+     }
    });

    contextMenu.querySelector("#sa-flag-menu-turbo").addEventListener("click", () => {
      closeContextMenu();
      greenFlag.dispatchEvent(new MouseEvent("click", { bubbles: true, shiftKey: true }));
    });
    contextMenu.querySelector("#sa-flag-menu-fps").addEventListener("click", () => {
      closeContextMenu();
      greenFlag.dispatchEvent(new MouseEvent("click", { bubbles: true, altKey: true }));
    });
    contextMenu.querySelector("#sa-flag-menu-mute").addEventListener("click", () => {
      closeContextMenu();
      greenFlag.dispatchEvent(new MouseEvent("click", { bubbles: true, ctrlKey: true }));
    });

    contextMenu.style.visibility = "hidden"; // Setting visibility here fixes a visual glitch on dynamic enable
    addon.tab.displayNoneWhileDisabled(contextMenu);
    addon.self.addEventListener("disabled", closeContextMenu);

    addon.tab.traps.vm.on("TURBO_MODE_ON", () => {
      contextMenu.querySelector("#sa-flag-menu-turbo").textContent = msg("turbo-off");
    });
    addon.tab.traps.vm.on("TURBO_MODE_OFF", () => {
      contextMenu.querySelector("#sa-flag-menu-turbo").textContent = msg("turbo-on");
    });

    while (true) {
      greenFlag = await addon.tab.waitForElement("[class^='green-flag_green-flag']", {
        markAsSeen: true,
        reduxEvents: ["scratch-gui/mode/SET_PLAYER", "fontsLoaded/SET_FONTS_LOADED", "scratch-gui/locales/SELECT_LOCALE"],
      });

      greenFlag.parentElement.appendChild(contextMenu);
      greenFlag.addEventListener("contextmenu", (e) => {
        if (addon.self.disabled) return;
        e.preventDefault();
+       contextMenuOpen = true;
        contextMenu.classList.add("sa-flag-menu-open");
        contextMenu.style.left = e.clientX + "px";
        contextMenu.style.top = e.clientY + "px";
      });
    }
  }
```

</details>
